### PR TITLE
Mesh cap update 

### PR DESCRIPTION
**Pull request by DeniseWorthen**
# Pull Request Summary

Update mesh cap with changes needed for eventual merge to develop.

## Description

- Allows users of the mesh cap to set runtime options for custom restart and history file names and to create netcdf output for gridded fields at run time. For UWM, these can be set in the nems.configure using the settings

```
  gridded_netcdfout = true
  user_sets_histname = true
  user_sets_restname = true
```

History file names will then be ``casename.ww3.hi.YYYY-MM-DD-SSSS.nc`` and ``casename.ww3.r.YYYY-MM-DD-SSSS`` where ``casename`` is set from nems.configure for each application (e.g ``ufs.hafs``)

- Replaces previous ``set_shel_inp`` and ``read_shel_inp`` routines with the applicable section from ``ww3_shel.F90`` to read either a inp or nml file. When used in UFS, the only required fields in the nml file are the list of forcing fields, the desired output fields, the name of the points file (if desired) and the stride for history and restart files. For example,

```
&input_nml
  input%forcing%winds = 'C'
  input%forcing%currents  = 'F'
  input%forcing%ice_conc  = 'F'
/
&output_type_nml
  type%field%list     = 'WND HS FP DP PHS PTP PDIR'
  type%point%file     = 'points.list'
/

&output_date_nml
  date%field%outffile  = '1'
  date%field%stride    = '3600'
  date%point%stride    = '3600'
  date%restart2%stride = '3600'
/
```

- Removes ``wav_tauice`` fields which are unused when coupling WW3 to CICE6.

### Issue(s) addressed
- https://github.com/NOAA-EMC/WW3/issues/679
- https://github.com/NOAA-EMC/WW3/issues/674
- https://github.com/NOAA-EMC/WW3/issues/673
- https://github.com/NOAA-EMC/WW3/issues/676

### Commit Message

Update mesh cap for custom user filenames, netcdf gridded output and logical control of some ifdef options.

co-authors : @mvertens 

### Testing

All UFS-WM baselines were run using hera.intel and hera.gnu and were B4B.

### CODE REVIEW

File: model/src/constants.F90 [hunk 1]
```diff
@@ -20,7 +20,7 @@ MODULE CONSTANTS
 !/
 !/    Copyright 2009-2012 National Weather Service (NWS),
 !/       National Oceanic and Atmospheric Administration.  All rights
-!/       reserved.  WAVEWATCH III is a trademark of the NWS. 
+!/       reserved.  WAVEWATCH III is a trademark of the NWS.
```
Comment (JessicaMeixner-NOAA) [inline model/src/constants.F90]: This entire file is just white space changes.  Should this go ahead and be reverted?

File: model/src/w3fldsmd.F90 [hunk 2]
```diff
@@ -25,10 +25,10 @@ MODULE W3FLDSMD
 !/    06-Dec-2010 : Change from GLOBAL (logical) to ICLO (integer) to
 !/                  specify index closure for a grid.   ( version 3.14 )
 !/                  (T. J. Campbell, NRL)
-!/    30-Oct-2012 : Implement tidal analysis            ( version 4.08 ) 
+!/    30-Oct-2012 : Implement tidal analysis            ( version 4.08 )
```
Comment (JessicaMeixner-NOAA) [inline model/src/w3fldsmd.F90]: This entire file is just white space changes.  Should this file be reverted to it's original?

File: model/src/w3servmd.F90 [hunk 3]
```diff
@@ -10,9 +10,9 @@ MODULE W3SERVMD
 !/                  +-----------------------------------+
 !/
 !/    For update log see individual subroutines.
-!/    12-Jun-2012 : Add /RTD option or rotated grid option. 
+!/    12-Jun-2012 : Add /RTD option or rotated grid option.
```
Comment (JessicaMeixner-NOAA) [inline model/src/w3servmd.F90]: Most changes are white space or style and are to me seem unrelated to the PR.   I'm wondering if it's worth keeping these changes in the PR or if it would be better to revert this file to its original state which would help keep this branch and develop in sync.

File: model/src/w3adatmd.F90 [hunk 4]
```diff
@@ -340,6 +340,9 @@ MODULE W3ADATMD
 !  7. Source code :
 !
 !/ ------------------------------------------------------------------- /
+      ! module default
```
Comment (JessicaMeixner-NOAA) [inline model/src/w3adatmd.F90]: Most changes are white space or style and to me seem unrelated to the PR.   I'm wondering if it's worth keeping these changes in the PR or if it would be better to revert this file to its original state which would help keep this branch and develop in sync.

File: model/src/w3gdatmd.F90 [hunk 5]
```diff
@@ -46,28 +46,28 @@ MODULE W3GDATMD
 !/    30-Oct-2009 : Implement curvilinear grid type.    ( version 3.14 )
 !/                  (W. E. Rogers & T. J. Campbell, NRL)
 !/    29-Oct-2010 : Implement unstructured grids        ( version 3.14.1 )
-!/                  (A. Roland and F. Ardhuin) 
+!/                  (A. Roland and F. Ardhuin)
```
Comment (JessicaMeixner-NOAA) [inline model/src/w3gdatmd.F90]: Most changes are white space or style and to me seem unrelated to the PR.   I'm wondering if it's worth keeping these changes in the PR or if it would be better to revert this file to its original state which would help keep this branch and develop in sync.

File: model/src/w3wavemd.F90 [hunk 6]
```diff
@@ -697,24 +697,20 @@ SUBROUTINE W3WAVE ( IMOD, ODAT, TEND, STAMP, NO_OUT &
...
-#ifdef W3_SBS
-      NDSOFLG = 99
-#endif
-#ifdef W3_MPI
       SBSED = .FALSE.
-#endif
-#ifdef W3_SBS
-      SBSED = .TRUE.
-#endif
+      if (w3_sbs_flag) then
```
Comment (JessicaMeixner-NOAA) [inline model/src/w3wavemd.F90:710]: Can this be kept as #ifdef W3_SBS and keep changing any #if-defs to a seperate PR?

File: model/src/wav_shr_flags.F90 [hunk 7]
```diff
@@ -0,0 +1,1201 @@
+!> @file wav_shr_flags
```
Comment (JessicaMeixner-NOAA) [inline model/src/wav_shr_flags.F90:1]: This seems related to if-def removals and not this PR, but perhaps I've missed something?  Is this file necessary for this PR?

**Review by JessicaMeixner-NOAA — CHANGES_REQUESTED** [review]
The standalone WW3 regression tests should be finishing soon.  These tests will help ensure there are no unintended impacts in the updates in w3wave (which visually all look good so far).  

For issue  #679 @MatthewMasarik-NOAA and I will work to extend the work here to remove duplicated code and have ww3_shel exes use the same separated subroutine for reading ww3_shel.inp. This issue will remain open after this PR. 

For  issue #674 I will be working as I mentioned the other day to extend this work so that the restart file writing will be variables read in from WW3 input files.   This will either change the reading form the nems.configure or allow for the nems.configure to over-ride this.  This issue will remain open after this PR. 

For issue issue #673  Work will be ongoing to extend this capability to reach the desired capabilities mentioned in the issues. 

I do believe this PR will close issue #678  which was not mentioned in the PR. 

I will report on the standalone WW3 regression tests as soon as the script has finished running, but wanted to provide this update now.  Apologies for the delay in review, we'll be working to reduce the review time for PRs in the future and thank you so much for the work you have contributed here!

File: model/src/constants.F90 (hunk 1 above)
Comment (DeniseWorthen) [reply model/src/constants.F90]: This is stripping of trailing white space. Trailing whitespace should not be committed.

File: model/src/w3fldsmd.F90 (hunk 2 above)
Comment (DeniseWorthen) [reply model/src/w3fldsmd.F90]: Again, this is stripping of trailing whitespace....same with the following white space changes.

File: model/src/w3wavemd.F90 (hunk 6 above)
Comment (DeniseWorthen) [reply model/src/w3wavemd.F90:710]: The issue here is that the sbs if-def is one of those "branching ifdefs" which are very difficult to code around w/o nesting additional ifdefs, which we wanted to avoid.

File: model/src/wav_shr_flags.F90 (hunk 7 above)
Comment (DeniseWorthen) [reply model/src/wav_shr_flags.F90:1]: We don't need all the flags, but it was simply easier to keep the original list together.

File: model/src/w3adatmd.F90 (hunk 4 above)
Comment (DeniseWorthen) [reply model/src/w3adatmd.F90]: implicit none should always be at the module level. Reverting this to the previous would mean reverting to a poor coding practice.

File: model/src/wav_shr_flags.F90 (hunk 7 above)
Comment (JessicaMeixner-NOAA) [reply model/src/wav_shr_flags.F90:1]: If we backed out the W3_SBS change, can all of this be backed out?

File: model/src/w3wavemd.F90 (hunk 6 above)
Comment (JessicaMeixner-NOAA) [reply model/src/w3wavemd.F90:710]: It'd be great to back this out if possible to keep different development separate.

File: model/src/w3adatmd.F90 (hunk 4 above)
Comment (JessicaMeixner-NOAA) [reply model/src/w3adatmd.F90]: From a code management perspective, it's preferred to keep the divergence from the develop branch to a minimum and keep the PR focused as possible.

File: model/src/constants.F90 (hunk 1 above)
Comment (JessicaMeixner-NOAA) [reply model/src/constants.F90]: The preference is to take care of trailing white spaces in it's own PR to the develop branch instead of being combined here.  While github does have a great feature to ignore white spaces in a review, it's still preferred to handle this separately.

**Comment by JessicaMeixner-NOAA** [comment]
The WW3 standalone tests performed as expected.

File: model/src/w3wavemd.F90 (hunk 6 above)
Comment (DeniseWorthen) [reply model/src/w3wavemd.F90:710]: No, it is not really possible to back out the change to this branching ifdef. I believe it would work to remove the wav_shr_flag file and put in a local definition of this flag.

Comment (JessicaMeixner-NOAA) [reply model/src/w3wavemd.F90:710]: I think that sounds like a great path forward!

Comment (DeniseWorthen) [reply model/src/w3wavemd.F90:710]: It turns out I cannot remove wav_shr_flags completely since it is also used by the mesh cap code. I removed it from w3wavemd.F90 and added locally defined flags.

Comment (JessicaMeixner-NOAA) [reply model/src/w3wavemd.F90:710]: Thanks @DeniseWorthen

**Comment by JessicaMeixner-NOAA** [comment]
@DeniseWorthen I really appreciate your updates to this branch.  I'll be re-running the standalone WW3 regression tests but expect no issues there.

**Comment by DeniseWorthen** [comment]
@JessicaMeixner-NOAA @MatthewMasarik-NOAA I had fixed the CI previously for the ufs branch but it is now failing. So I checked out develop and pushed a copy to my fork and it also fails. I don't know why---can one of you figure it out? The branch I created and tested is [testci](https://github.com/DeniseWorthen/WW3/tree/testci). You can see the workflow error in the Intel Linux Build / setup (push)

```
Note that versions of Spack older than 0.19 may not be able to use the updated configuration.
==> Error: list index out of range
Error: Process completed with exit code 3.
```

**Comment by JessicaMeixner-NOAA** [comment]
@DeniseWorthen yes, we can look at the ci for this branch and develop.  Realistically, we will not be able to get it fixed before this branch's turn in the ufs-queue.

**Comment by JessicaMeixner-NOAA** [comment]
Results of standalone regression tests were as expected.

**Review by JessicaMeixner-NOAA — APPROVED** [review]
Updates were made based on review and ufs-waether-model regression tests passed (https://github.com/ufs-community/ufs-weather-model/pull/1366) in addition to ww3 standalone.  

Thanks @DeniseWorthen !
